### PR TITLE
Fixed AttributeError during pre_migrate when permission constraints are applied to custom field.

### LINKED ISSUE #6081: Permission constraint applied to Custom Field causes pre_migrate KeyTransform AttributeError exception
### Environment
* Nautobot version (Docker tag too if applicable): tag release `v1.6.26`, Docker environment
* Python version: 3.8
* Database platform, version: PG15
* Middleware(s): N/A

The Permission and its constraints are working as expected in Nautobot, the issue is the `pre_migrate` management command throws an AttributeError exception at this [line](https://github.com/nautobot/nautobot/blob/28d12b24113e422fc7a4a33b3acad7d605cda07c/nautobot/core/management/commands/pre_migrate.py#L379) when we have the Custom Field permission constraint applied.

### Steps to Reproduce
1. Create a Custom Field for DCIM Device objects, eg: `test_custom_field`
2. Create a Permission (with Constraints) for View to DCIM Device objects, eg:
```json
{"_custom_field_data__test_custom_field__in": ["test-1", "test-2"]}
```
image

3. (Optional) Add the Custom Field data required to a Device, eg: "test-1" to `test_custom_field` to observe filter working as expected
4. Run `$ nautobot_server pre_migrate` management command, reproducing the error

### Expected Behavior
pre_migrate management command completes as normal

### Observed Behavior

```bash
root@docker:/source# nautobot-server pre_migrate
>>> Running check: check_configcontext_uniqueness...
>>> Running check: check_exporttemplate_uniqueness...
>>> Running check: check_virtualchassis_uniqueness...
>>> Running interface/vminterface vrf uniqueness checks...
>>> Running permission constraint checks...
Traceback (most recent call last):
  File "/usr/local/bin/nautobot-server", line 6, in <module>
    sys.exit(main())
  File "/source/nautobot/core/cli.py", line 54, in main
    run_app(
  File "/source/nautobot/core/runner/runner.py", line 297, in run_app
    management.execute_from_command_line([runner_name, command] + command_args)
  File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/source/nautobot/core/management/commands/pre_migrate.py", line 457, in handle
    check_permissions_constraints(self)
  File "/source/nautobot/core/management/commands/pre_migrate.py", line 379, in check_permissions_constraints
    related_model = lookup.lhs.target.related_model or lookup.lhs.target.model
AttributeError: 'KeyTransform' object has no attribute 'target'
```

### Note:
I have explored alternative variations for the permission constraint to no avail, eg:
```json
{"cf_interface_test_custom_field__in": ["test-1", "test-2"]}
```
However the permission constraint cannot be applied and indicates I should use `_custom_field_data`
image

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes #6081 
# What's Changed

# Screenshots

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/core/tests/test_commands.py [hunk 1]
```diff
@@ -80,3 +96,22 @@ def test_virtualchassis_failure(self):
 
         with self.assertRaises(CommandError):
             self.run_command()
+
+    def test_permission_constraints_failure(self):
+        """Test that duplicate VirtualChassis objects result in a failure."""
```
Comment (glennmatthews) [inline nautobot/core/tests/test_commands.py:101]: Wrong docstring?

**Review by glennmatthews — APPROVED** [review]
Besides the linting issue, this looks good - thanks for the test case!

File: nautobot/core/tests/test_commands.py [hunk 2]
```diff
@@ -80,3 +95,22 @@ def test_virtualchassis_failure(self):
...
```
Comment (timizuoebideri1) [inline nautobot/core/tests/test_commands.py:100]:
```suggestion
        """Test permission constraints logs warning when needed for CustomField."""
```

File: nautobot/core/tests/test_commands.py [hunk 3]
```diff
@@ -80,3 +95,22 @@ def test_virtualchassis_failure(self):
 
         with self.assertRaises(CommandError):
             self.run_command()
+
+    def test_permission_constraints_failure(self):
+        """Test permission constraints logs warning when needed for CustomField."""
+        device_ct = ContentType.objects.get_for_model(Device)
+        site_ct = ContentType.objects.get_for_model(Device)
+        custom_field = CustomField.objects.create(type=CustomFieldTypeChoices.TYPE_TEXT, name="custom_field")
+        custom_field.content_types.add(site_ct)
```
Comment (HanlinMiao) [inline nautobot/core/tests/test_commands.py:104]: I am a little confused here. Do we need line 102 and line 104 here?
We are trying to add `site_ct` which is actually the ContentType for Device?
The warning should be raised just by including `site` in the constraints of `obj_perm` right?

Comment (glennmatthews) [reply nautobot/core/tests/test_commands.py:104]: Good catches. Definitely 102 should be `get_for_model(Site)` if anything. But it'd be interesting to remove lines 102 through 104 and see if that affects the test at all, especially given that the custom_field has name `custom_field` but the obj_perm is using `custom_field_in` (probably meant to be `custom_field__in` instead?)...

File: nautobot/core/management/commands/pre_migrate.py [hunk 4]
```diff
@@ -376,42 +377,61 @@ def check_permissions_constraints(command):
...
                         )
 
-                if lookup.lhs.field.name == "slug" and related_model is not GitRepository:
-                    warnings.append(
-                        f"ObjectPermission '{perm}' (id: {perm.id}) has a constraint that references "
-                        f"a 'slug' field that will be deleted by the Nautobot 2.0 migration.\n"
-                        + json.dumps(perm.constraints, indent=4)
-                    )
+                    if related_model in field_change_models:
+                        if field_name in field_change_models[related_model]:
```
Comment (glennmatthews) [inline nautobot/core/management/commands/pre_migrate.py:411]: Is `field_name` (`{col.field.name}{suffix}`) correct here or should it be just `col.field.name`?

File: nautobot/core/tests/test_commands.py [hunk 5]
```diff
@@ -25,6 +29,17 @@ def setUp(self):
...
+        # Adding this test data to assert that https://github.com/nautobot/nautobot/issues/6081 is fixed
+        ct = ContentType.objects.get_for_model(GitRepository)
+        custom_field = CustomField.objects.create(type=CustomFieldTypeChoices.TYPE_TEXT, name="test_custom_field")
+        custom_field.content_types.add(ct)
+        obj_perm = ObjectPermission.objects.create(
+            name="Test permission",
+            constraints={"_custom_field_data__test_custom_field__in": ["test-1", "test-2"]},
+            actions=["view"],
+        )
+        obj_perm.object_types.add(ct)
```
Comment (glennmatthews) [inline nautobot/core/tests/test_commands.py:41]: Should there be a test somewhere that checks for a warning to be logged for `Test permission`, independent of the `Test permission 2` test below?

File: nautobot/core/tests/test_commands.py [hunk 6]
```diff
@@ -57,6 +58,11 @@ def test_success(self):
         out, err = self.run_command()
 
         self.assertIn("All pre-migration checks passed.", out)
+        # Assert Permission constrain warning not logged
+        self.assertNotIn(
+            f"ObjectPermission '{self.obj_perm.name}' (id: {self.obj_perm.pk}) has a constraint that references a model (nautobot.extras.models.datasources.GitRepository) that will be migrated to a new model by the Nautobot 2.0 migration.",
+            out,
+        )
```
Comment (glennmatthews) [inline nautobot/core/tests/test_commands.py:65]: This seems like a really specific string for a negative test. Should it instead just be something like `self.assertNotIn(f"ObjectPermission '{self.obj_perm.name}'", out)` to catch *any* unexpected message for this permission?

File: nautobot/core/tests/test_commands.py [hunk 7]
```diff
@@ -42,6 +58,11 @@ def test_success(self):
         out, err = self.run_command()
 
         self.assertIn("All pre-migration checks passed.", out)
+        # Assert Permission constrain warning not logged
+        self.assertNotIn(
+            f"ObjectPermission '{self.obj_perm.name}' (id: {self.obj_perm.pk}) has a constraint that references a model (nautobot.extras.models.datasources.GitRepository) that will be migrated to a new model by the Nautobot 2.0 migration.",
```
Comment (timizuoebideri1) [inline nautobot/core/tests/test_commands.py:63]:
```suggestion
            f"ObjectPermission '{self.obj_perm.name}'",
```

File: nautobot/core/tests/test_commands.py [hunk 8]
```diff
@@ -42,6 +58,11 @@ def test_success(self):
         out, err = self.run_command()
 
         self.assertIn("All pre-migration checks passed.", out)
+        # Assert Permission constrain warning not logged
+        self.assertNotIn(
+            f"ObjectPermission '{self.obj_perm.name}'",
+            out,
+        )
```
Comment (timizuoebideri1) [inline nautobot/core/tests/test_commands.py:65]:
```suggestion
        self.assertNotIn(f"ObjectPermission '{self.obj_perm.name}'",  out)
```

**Pull request merged**